### PR TITLE
test: backfill extraction and telemetry parity coverage

### PR DESCRIPTION
**Pull request by yczhang-nv**
#### Overview

Backfill regression and parity coverage for the shared extraction seams established by #291, #300, #301, #304, and #318, so agent-payload extraction, provider-schema normalization, and exporter projection cannot drift silently. Test-only: the only `src` edits are `#[cfg(test)]` module wiring for two new suites; no production behavior changes.

- [x] I confirm this contribution is my own work, or I have the right to submit it under this project's license.
- [x] I searched existing issues and open pull requests, and this does not duplicate existing work.

#### Details

**Agent payload extraction** (`crates/cli/tests/coverage/adapters_tests.rs`, 15 tests):

- Per-host partial-sparse payloads for Claude Code, Codex, and Hermes (real identifiers kept, absent fields synthesized or null at the adapter boundary).
- Path-precedence fallback-chain walks for session IDs, event names, subagent IDs, tool-call IDs, and tool results/status.
- Hermes tool result and status extraction (previously uncovered), including explicit-status vs event-name-derived status interaction.
- Claude Code LLM-hint extraction and hint field precedence chains.
- JSON-path primitive edge cases: empty-string filtering, array intermediates, deep nesting, all-paths-miss.

**Provider request extraction** (`crates/cli/tests/coverage/alignment_tests.rs`, 6 tests):

- First coverage for `request_affinity_key` (route-gated, length-bounded) and `gateway_turn_input` (Claude Code + Anthropic Messages only).
- Header-only session-id routing on the Models and Count-Tokens routes, contrasted with the body fallbacks on the Responses/Chat routes.
- Gateway route name round-trips.

**Codec parity** (`crates/core/tests/unit/codec/parity_tests.rs`, new, 14 tests):

Each test builds the same logical scenario in all three provider schemas (OpenAI Chat Completions, Anthropic Messages, OpenAI Responses) and asserts the normalized output agrees: model name, finish reasons, tool calls (full-struct equality), usage incl. cache-read tokens, provider-reported and catalog-estimated cost, hint hardening (`normalize_request_with_hint`), and request normalization. Schema-inherent divergences are asserted explicitly as part of the parity contract, e.g. Responses normalizes a tool-call turn to `Complete` (no tool-use terminal status), cache-write tokens are Anthropic-only, reasoning tokens are Responses-only.

**Exporter parity** (`crates/core/tests/unit/observability/exporter_parity_tests.rs`, new, 9 tests):

A shared harness feeds one event stream to the ATIF exporter, `OpenTelemetrySubscriber`, and `OpenInferenceSubscriber` (in-memory span exporters) and asserts the projected facts agree: cost totals, usage, model names, tool-call projection, replay-payload preservation, and the consolidated manual-fallback path — run against both OpenAI-Chat-shaped and Anthropic-shaped payloads. Intentional projection divergences are pinned with explicit assertions so drift fails loudly, notably: OTel emits no token-count attributes, reasoning facts are ATIF-only, ATIF sums cache read+write while OpenInference splits them, and OpenInference omits raw LLM request JSON. Cost-currency policy boundaries stay covered by the existing per-exporter tests from #304 and are not duplicated here.

**Validation**

- `just test-rust` — full workspace green
- `cargo clippy --workspace --all-targets -- -D warnings` — clean
- `cargo fmt --all --check` — clean
- `uv run pre-commit run --files <changed files>` — all hooks green (SPDX, fmt, clippy, check, linkcheck)
- Rebased onto current `main` and re-ran `cargo test -p nemo-relay` and `cargo test -p nemo-relay-cli` — green, including the parity suites against the latest observability changes
- Binding matrix not run: the change is test-only; the `#[cfg(test)]` modules are not compiled into the library targets, so binding-facing behavior is unchanged

**Breaking changes**

None.

#### Where should the reviewer start?

`crates/core/tests/unit/observability/exporter_parity_tests.rs` — the `export_through_all_exporters` harness and the explicitly pinned divergences. These assertions encode current behavior. The OTel cost-only projection and the ATIF cache-sum vs OpenInference cache-split asymmetry match the token/cost field-semantics contract documented in #330; the remaining pinned divergences (reasoning projection being ATIF-only, OTel not flattening LLM tool calls, OpenInference omitting raw LLM request JSON) encode current behavior without a recorded decision — if any is ruled a defect, the fix should land together with flipping the pinning assertion.

Then `crates/core/tests/unit/codec/parity_tests.rs` for the cross-schema normalization contract, and the adapter/alignment additions in `crates/cli/tests/coverage/`.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- Relates to: none (test backfill for the extraction-strategy refactors #291, #300, #301, #304, #318)

## Summary by CodeRabbit

* **Tests**
  * Added coverage for Hermes and Claude tool adapter/extractor precedence, partial payload handling, and hint/session/model/request-id resolution (including null/empty-string and fallback-chain behaviors).
  * Expanded gateway alignment tests for affinity-key/session-id gating and route/provider-specific prompt building.
  * Introduced core codec and observability cross-provider/exporter parity tests validating normalized fields, tool-call mapping, and consistent usage/cost/model/replay projections (with expected divergence checks).

### CODE REVIEW

File: crates/cli/tests/coverage/adapters_tests.rs [hunk 1]
```diff
@@ -1072,3 +1072,529 @@ fn stop_responses_preserve_vendor_shapes() {
...
+            Some(expected)
+        );
+        payload.as_object_mut().unwrap().remove(key);
+    }
+    assert_eq!(
+        CODEX_PAYLOAD_EXTRACTOR.subagent_id(&payload, &headers),
+        None
+    );
+}
+
```
[the remark this answers is not in the document]
Comment (yczhang-nv) [reply crates/cli/tests/coverage/adapters_tests.rs:1495]: Extracted an `assert_string_fallback_chain` helper (payload, chain tuples, extraction closure) and converted all seven chain-walk sites to it, including the two llm-hint chains not listed in the comment. The tool-result walk maps its `Value` result through a string expectation so the same helper covers it.

File: crates/core/tests/unit/codec/parity_tests.rs [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+    //   + 500 completion * 0.60/M + 200 cache-read * 0.075/M
+    assert_eq!(chat_cost.total, Some(0.000_435));
+    assert_eq!(chat_cost.currency, "USD");
+    assert_eq!(chat_cost.source, CostSource::ModelPricing);
+    assert_eq!(chat_cost.pricing_provider.as_deref(), Some("test"));
+    assert_eq!(
+        chat_cost.pricing_model.as_deref(),
+        Some("parity-priced-model")
+    );
+}
```
[the remark this answers is not in the document]
Comment (yczhang-nv) [reply crates/core/tests/unit/codec/parity_tests.rs:548]: Switched the hand-derived literal to a tolerance comparison (`abs < 1e-9`). The cross-codec `assert_eq!(chat_cost, anthropic_cost)` full-struct comparisons stay exact deliberately: all three codecs share one estimation code path, so bit-identical output across codecs is the parity contract.

File: crates/core/tests/unit/observability/exporter_parity_tests.rs [hunk 3]
```diff
@@ -0,0 +1,803 @@
...
+    attributes
+        .iter()
+        .map(|attribute| {
+            (
+                attribute.key.as_str().to_string(),
+                attribute.value.to_string(),
+            )
+        })
+        .collect()
+}
```
[the remark this answers is not in the document]
Comment (yczhang-nv) [reply crates/core/tests/unit/observability/exporter_parity_tests.rs:102]: `attr_map` now builds the map with explicit inserts and panics on any duplicate attribute key, so a duplicate-emitting regression fails loudly instead of being collapsed.

File: crates/core/tests/unit/codec/parity_tests.rs [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+        assert_eq!(cost.input, None);
+        assert_eq!(cost.output, None);
+    }
+}
+
+#[test]
+fn test_estimated_cost_parity_for_identical_model_and_usage() {
+    let _pricing_guard = pricing_test_mutex().lock().unwrap();
+    install_parity_pricing("parity-priced-model");
+    let _reset_guard = ResetPricingResolverGuard;
```
[the remark this answers is not in the document]
Comment (yczhang-nv) [reply crates/core/tests/unit/codec/parity_tests.rs]: Applied to all four pricing-mutex lock sites in the new suites (this one plus three in exporter_parity_tests.rs), recovering the guard via `into_inner`. Recovery is safe here because every test reinstalls the pricing resolver and resets it on drop, so a poisoned lock never exposes stale state.

**Review by willkill07 — APPROVED** [review]
Codex, Coderabbit, and I have reached consensus that this is good to go!

Only noteworthy comments:
- exporter parity tests cover OpenAI Chat, Anthropic, and manual fallback—but not provider-shaped OpenAI Responses payloads
- Tool-call exporter coverage is also Chat-only, so Responses function_call and Anthropic tool_use aren’t exercised through the exporter pipeline

I don't think it needs to be addressed prior to merge.

**Review by zhongxuanwang-nv — APPROVED** [review]
Used powerful AIs, and I think nothing more is needed to be addressed!

**Review by mnajafian-nv — APPROVED** [review]
LGTM!
